core/micro-services/workflow-compiling-service/src/main/scala/edu/uci/ics/amber/compiler/WorkflowCompiler.scala

```diff
@@ -0,0 +1,183 @@
+
+        val subPlan = logicalOp.getPhysicalPlan(context.workflowId, context.executionId)
+        subPlan
+          .topologicalIterator()
+          .map(subPlan.getOperator)
+          .foreach({ physicalOp =>
+            {
+              val externalLinks = logicalPlan
+                .getUpstreamLinks(logicalOp.operatorIdentifier)
+                .filter(link => physicalOp.inputPorts.contains(link.toPortId))
+                .flatMap { link =>
+                  physicalPlan
+                    .getPhysicalOpsOfLogicalOp(link.fromOpId)
+                    .find(_.outputPorts.contains(link.fromPortId))
+                    .map(fromOp =>
+                      PhysicalLink(fromOp.id, link.fromPortId, physicalOp.id, link.toPortId)
+                    )
+                }
+
+              val internalLinks = subPlan.getUpstreamPhysicalLinks(physicalOp.id)
+
+              // Add the operator to the physical plan
+              physicalPlan = physicalPlan.addOperator(physicalOp.propagateSchema())
+
+              // Add all the links to the physical plan
+              physicalPlan = (externalLinks ++ internalLinks)
+                .foldLeft(physicalPlan) { (plan, link) => plan.addLink(link) }
+            }
+          })
+      } match {
+        case Success(_) =>
+        case Failure(err) =>
+          errorList match {
+            case Some(list) => list.append((logicalOpId, err))
+            case None       => throw err
+          }
+      }
+    )
+    physicalPlan
+  }
+
+  /**
+    * Compile a workflow to physical plan, along with the schema propagation result and error(if any)
+    *
+    * @param logicalPlanPojo the pojo parsed from workflow str provided by user
+    * @return WorkflowCompilationResult, containing the physical plan, input schemas per op and error per op
+    */
+  def compile(
+      logicalPlanPojo: LogicalPlanPojo
+  ): WorkflowCompilationResult = {
+    val errorList = new ArrayBuffer[(OperatorIdentity, Throwable)]()
+    var opIdToInputSchema: Map[OperatorIdentity, List[Option[Schema]]] = Map()
+    // 1. convert the pojo to logical plan
+    var logicalPlan: LogicalPlan = LogicalPlan(logicalPlanPojo)
+
+    // 2. Manipulate logical plan by:
+    // - inject sink
+    logicalPlan = SinkInjectionTransformer.transform(
+      logicalPlanPojo.opsToViewResult,
+      logicalPlan
+    )
+    // - resolve the file name in each scan source operator
+    logicalPlan.resolveScanSourceOpFileName(Some(errorList))
+
+    // 3. expand the logical plan to the physical plan,
+    val physicalPlan = expandLogicalPlan(logicalPlan, Some(errorList))
+    if (errorList.isEmpty) {
+      // no error during the expansion, then do:
+      // - collect the input schema for each op
+      opIdToInputSchema = collectInputSchemaFromPhysicalPlan(physicalPlan)
+      // - check for static errors
+      logicalPlan.propagateWorkflowSchema(context, Some(errorList))
+    }
+
+    WorkflowCompilationResult(
```

Comment:
I still hope you can use two messages to handle it. One for success and one for failure. But we can do it later.